### PR TITLE
Raise an error when @fold contains no @outputs (#16) 

### PR DESCRIPTION
**Pull request by benlongo**
Covers #16. It seems that this already raises an error.

### CODE REVIEW

**Comment by obi1kenobi** [comment]
Hi @benlongo, thank you for the PR!

Unfortunately, I think the raised error is not the correct one -- I think the fact that the *entire query* is missing `@output` fields is causing an error, and not the `@fold` scope specifically. I think you'll be able to confirm this with the following example:
```
{
    Animal {
        uuid @output(out_name: "uuid")
        out_Animal_ParentOf @fold {
            name
        }
    }
}
```

**Comment by benlongo** [comment]
Hello @obi1kenobi 
Your test case does indeed work, my mistake. 
As a side note, is it desirable that the tests themselves are captured within the coverage report?

**Comment by obi1kenobi** [comment]
If you'd like to add your example query to the `test_no_output_fields` test near line 697, I'd be happy to merge it -- since it's another nice example of a query that has no output fields.

I'd also love to help you address #16. The fix should be part of the front-end of the compiler (`compiler_frontend.py`), in two parts. I'd suggest that we modify `_compile_property_ast` to add the names of outputs to the fold scope (if in one). Then, `_compile_vertex_ast` on exit from a `@fold` scope (near [line 502 of that file](https://github.com/kensho-technologies/graphql-compiler/blob/master/graphql_compiler/compiler/compiler_frontend.py#L502) ) could check that the fold scope being exited has a non-empty list of outputs that were selected in that scope. This should work since `@fold` cannot exist within a `@fold` scope.

Let me know what you think!

Regarding the coverage report, I didn't mean to include the tests. I just haven't gotten around to removing them yet -- feel free to open an issue / PR if you'd like.

**Comment by benlongo** [comment]
Thanks for the direction! 
For storing the outputs in the fold context, should we modify `context['fold']` to contain something like `'outputs': []` and build that within `_compile_property_ast`? I see that `context['outputs'][output_name]` contains `'fold'` [(around 294) ](https://github.com/kensho-technologies/graphql-compiler/blob/master/graphql_compiler/compiler/compiler_frontend.py#L294) so it could be possible to just look through that and determine if there exists a fold with no outputs. 

I'll make another PR to ignore the test files for the coverage report.

**Comment by obi1kenobi** [comment]
Both approaches work: the first one is easier, but the second one is probably cleaner since it'll have no data denormalization (output fold data is only stored once) and may prevent bugs in the future. Let's go with that idea -- it shouldn't be too bad.

How about a function call on exit from the `@fold` scope near [line 502](https://github.com/kensho-technologies/graphql-compiler/blob/master/graphql_compiler/compiler/compiler_frontend.py#L502), to a function similar to:
```
def _validate_fold_has_outputs(fold_data, outputs):
    # At least one output in the outputs list must point to the fold_data,
    # or the scope corresponding to fold_data had no @outputs and is illegal.
    for output in outputs:
        if output['fold'] is fold_data:
            return True
```

Alternatively, if you don't want to use `is`, you could compare the `location` key of the fold -- it should be unique for each fold and supports `==`.

What do you think?

**Comment by benlongo** [comment]
I agree the second approach is cleaner, I had a go at it.

File: graphql_compiler/compiler/compiler_frontend.py [hunk 1]
```diff
@@ -525,6 +526,18 @@ def _compile_vertex_ast(schema, current_schema_type, ast,
...
 
 
+def _validate_fold_has_outputs(fold_data, outputs):
+    # At least one output in the outputs list must point to the fold_data,
+    # or the scope corresponding to fold_data had no @outputs and is illegal.
+    for output in outputs.values():
+        if output['fold'] is fold_data:
+            return True
+    
+    raise GraphQLCompilationError(u'A @fold must contain at least one @output!'
```
Comment (obi1kenobi) [inline graphql_compiler/compiler/compiler_frontend.py]: Nitpick: I'd expand the explanation here, in order to help out users. Maybe something like:
```
Each @fold scope must contain at least one field marked @output. Encountered a @fold with no outputs at query location: {}
```

File: graphql_compiler/compiler/compiler_frontend.py [hunk 2]
```diff
@@ -525,6 +526,19 @@ def _compile_vertex_ast(schema, current_schema_type, ast,
...
+    for output in outputs.values():
+        if output['fold'] is fold_data:
+            return True
+    
+    raise GraphQLCompilationError(u'Each @fold scope must contain at least one field '
+                                  u'marked @output. Encountered a @fold with no outputs '
+                                  u'at query location: {}'.format(fold_data['root']))
+
+
+
```
Comment (obi1kenobi) [inline graphql_compiler/compiler/compiler_frontend.py]: Hmm it seems that I forgot to add `flake8` to the project -- my bad! Anyway, the 3 consecutive blank lines don't pass flake8:
```
$ flake8 ./graphql_compiler/
./graphql_compiler/compiler/compiler_frontend.py:541:1: E303 too many blank lines (3)
```

**Comment by benlongo** [comment]
I didn't even bother to look for pep, figured the combination of pydocstyle graphql_compiler/

isort --check-only --verbose --recursive graphql_compiler/

pylint graphql_compiler/

bandit -r graphql_compiler/

**Pull request merged**